Tests/WebAuthnTests/WebAuthnManagerTests+Authentication.swift

```diff
@@ -0,0 +1,188 @@
+            expect: WebAuthnError.invalidClientDataJSON
+        )
+    }
+
+    func testFinishAuthenticationFailsIfClientDataJSONDecodingFails() throws {
+        try assertThrowsError(finishAuthentication(clientDataJSON: "abc")) { (_: DecodingError) in
+            return
+        }
+    }
+
+    func testFinishAuthenticationFailsIfAuthenticatorDataIsInvalid() throws {
+        try assertThrowsError(
+            finishAuthentication(authenticatorData: "%"),
+            expect: WebAuthnError.invalidAuthenticatorData
+        )
+    }
+
+    func testFinishAuthenticationFailsIfCeremonyTypeDoesNotMatch() throws {
+        var clientDataJSON = TestClientDataJSON()
+        clientDataJSON.type = "webauthn.create"
+        try assertThrowsError(
+            finishAuthentication(clientDataJSON: clientDataJSON.base64URLEncoded),
+            expect: CollectedClientData.CollectedClientDataVerifyError.ceremonyTypeDoesNotMatch
+        )
+    }
+
+    func testFinishAuthenticationFailsIfRelyingPartyIDHashDoesNotMatch() throws {
+        try assertThrowsError(
+            finishAuthentication(
+                authenticatorData: TestAuthDataBuilder()
+                    .validAuthenticationMock()
+                    .rpIDHash(fromRpID: "wrong-id.org")
+                    .buildAsBase64URLEncoded()
+            ),
+            expect: WebAuthnError.relyingPartyIDHashDoesNotMatch
+        )
+    }
+
+    func testFinishAuthenticationFailsIfUserPresentFlagIsNotSet() throws {
+        try assertThrowsError(
+            finishAuthentication(
+                authenticatorData: TestAuthDataBuilder()
+                    .validAuthenticationMock()
+                    .flags(0b10000000)
+                    .buildAsBase64URLEncoded()
+            ),
+            expect: WebAuthnError.userPresentFlagNotSet
+        )
+    }
+
+    func testFinishAuthenticationFailsIfUserIsNotVerified() throws {
+        try assertThrowsError(
+            finishAuthentication(
+                authenticatorData: TestAuthDataBuilder()
+                    .validAuthenticationMock()
+                    .flags(0b10000001)
+                    .buildAsBase64URLEncoded(),
+                requireUserVerification: true
+            ),
+            expect: WebAuthnError.userVerifiedFlagNotSet
+        )
+    }
+
+    func testFinishAuthenticationFailsIfCredentialCounterIsNotUpToDate() throws {
+        try assertThrowsError(
+            finishAuthentication(
+                authenticatorData: TestAuthDataBuilder()
+                    .validAuthenticationMock()
+                    .counter([0, 0, 0, 1]) // signCount = 1
+                    .buildAsBase64URLEncoded(),
+                credentialCurrentSignCount: 2
+            ),
+            expect: WebAuthnError.potentialReplayAttack
+        )
+    }
+// c155900296b95a46f75d57328e4523910e74d629331fcd12cce2e95644114182
+// cf28f38af10da2d9ca9e2d3bd373f0574253404075ca5d748869bbb01f8677be
```

Comment:
What do these mean?

Reply:
ah that's an artefact of debugging the signature verification :D Will remove that.